### ISSUE TITLE
LG-75: mpHealth 2.0 - Implementation

### ISSUE THREAD

**Issue by pgunapal** [OpenLiberty/open-liberty#7415]
Eclipse MicroProfile Health Check: https://github.com/eclipse/microprofile-health/tree/2.0.1

Aha: https://bigblue.aha.io/features/API-32
Aha (Spec/API/TCK): https://bigblue.aha.io/features/LG-74
Aha (Implementation): https://bigblue.aha.io/features/LG-75

ID: https://github.ibm.com/was-liberty/liberty-docs/issues/1311
OpenLiberty Guides Updates: https://github.com/OpenLiberty/guides-common/issues/289

**Comment by pgunapal** [comment]
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
 ## List of Steps to complete or get approvals / sign-offs for Onboarding to the Liberty release (GM date)
 
Instructions:
 - Do the actions below and mark them complete in the checklist when they are done.
 - Make sure all feature readiness approvers put the appropriate tag on the epic to indicate their approval.
 
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
 ## TARGET COMPLETION DATE **Before Development Starts or 8 weeks before Onboarding** 
- [ ] POC Design / WAD Review Scheduled (David Chang) or N/A. AHA Feature marked.
- [ ] POC Design / WAD Reviewed (Feature Owner) or N/A. AHA Feature marked.
- [ ] Complete any To-Do's from the POC Review, and create a new TO-DO in the AHA Feature for the "Design Approval Request" and assign it to Ian Robinson. Leave the date blank. (Feature Owner)
- [ ] Design / WAD Approval (Ian Robinson) or N/A. AHA Feature marked.
- [ ] No Design / No WAD Approval (Arthur De Magalhaes - cloud / Alasdair Nottingham - server) or N/A. AHA Feature marked.
- [ ] AHA Feature has link to Github epic or issue. Github epic has link back to AHA Feature. (Feature Owner or Backlog Subtribe PM).
- [ ] SVT Requirements identified. (Epic owner / Feature owner with SVT focal point)
- [ ] ID Requirements identified. (Epic owner / Feature owner with ID focal point)
- [ ] Create a child task of the epic entitled "FAT Approval Test Summary"
 
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
 ## TARGET COMPLETION DATE **3 weeks before Onboarding** 
- [ ] Identify all open source libraries that are changing or are new. Work with Legal Release Services (Cass Tucker or Release PM) to get open source cleared and approved. Or N/A. (Epic Owner).   New or changed open source impacts license and Certificate of Originality.
 
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
 ## TARGET COMPLETION DATE ** 3 weeks before Onboarding** 
- [ ] All new or changed PII messages are checked into the integration branch, before the last translation shipment out. (Epic Owner)
 
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
 ## TARGET COMPLETION DATE **2 weeks before Onboarding** 
- [ ] Implementation complete. (Epic owner / Feature owner) 
- [ ] All function tests complete. Ready for FAT Approval. (Epic owner / Feature owner)
- [ ] Review all known issues for Stop Ship. (Epic owner / Feature owner / PM)
 
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
 ## APPROVALS with TARGET COMPLETION DATE **2 to 1 week before Onboarding** 
Prereq: You must have the Design Approval or No Design Approval on the AHA Feature.
- [ ] **Accessibility** - (G Scott Johnston). Accessibility testing is complete or N/A. Approver adds label focalApproved:accessibility to the Epic in Github.
- [ ] **FAT Liberty SOE** - (Kevin Smith). SOE FATS are running successfully or N/A . Approver adds label focalApproved:fat to the Epic in Github.
- [ ] **Globalization** (Marika Joannidis - Liberty / Simy Cheeran - tWAS). Translation is complete or N/A. TVT - complete or N/A. Approver adds label focalApproved:globalization to the Epic in Github.
- [ ] **ID** - (Kareen Deen). Documentation work is complete or N/A . Approver adds label focalApproved:id to the Epic in Github.
- [ ] **Performance** - (Jared Anderson). Performance testing is complete with no high severity defects or N/A . Approver adds label focalApproved:performance to the Epic in Github.
- [ ] **Serviceability** - (Don Bourne). Serviceability has been addressed. 
- [ ] **STE** - (Swati Kasundra). STE chart deck is complete or N/A . Approver adds label focalApproved:ste to the Epic in Github.
- [ ] **SVT** - (Greg Ecock - Cloud, Brian Hanczaryk- APS). SVT is complete or N/A . Approver adds label focalApproved:svt to the Epic in Github.
 
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
## TARGET COMPLETION DATE **1 week before Onboarding** 
- [ ] No Stop Ship issues for the feature. (Epic owner / Feature owner / Release PM)
- [ ] Ship Readiness Review and Release Notes completed (Epic owner / Feature owner / Release PM)
- [ ] Github Epic and Epic's issues are closed / complete. All PRs are committed to the master branch. (Epic owner / Feature owner / Backlog Subtribe PM)
 
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
 ## NOT REQUIRED FOR A FEATURE
- [ ] **OL Guides** - (Yee-Kang Chang). Assessment for OL Guides is complete or N/A. 
- [ ] **WDT** - (Leonard Theivendra). WDT work complete or N/A.
 
 
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
## Related Deliverables TARGET COMPLETION DATE **General Availability** 
- [ ] Blog article writeup (Epic owner / Feature owner / Laura Cowen)

**Comment by gscottj** [comment]
This feature does not add any new or change any existing UI. No AVT is required. Accessibility focal approved.

**Comment by fmhwong** [comment]
WAD approved in https://bigblue.aha.io/features/LG-75

**Comment by fmhwong** [comment]
@jhanders34 There is no significant change in performance.  May I get your approval on performance?

**Comment by fmhwong** [comment]
@skasund Prashanth is currently out on vacation.  Can we get your approval on deferred STE?

**Comment by fmhwong** [comment]
@lauracowen GA blog post has been created.  May I get the ID approval?

**Comment by lauracowen** [comment]
> @lauracowen GA blog post has been created. May I get the ID approval?

You need @chirp1 for ID approval. On the backlog for 19.0.0.7, there are two issues that, when completed, should cover the docs work for this feature:
- https://app.zenhub.com/workspaces/docs-5c7031f61d4580192beb4070/issues/openliberty/docs/343 (background concept of MP Health)
- https://app.zenhub.com/workspaces/docs-5c7031f61d4580192beb4070/issues/openliberty/docs/308 (19.0.0.7-specific info about MP Health)

**Comment by lauracowen** [comment]
> @lauracowen GA blog post has been created. May I get the ID approval?

Do you have a link to the GA blog post, pls?

**Comment by fmhwong** [comment]
GA blog post https://github.com/openliberty/open-liberty/issues/7895

**Comment by donbourne** [comment]
**Serviceability Approval Comment** - Please answer the following questions for serviceability approval:

1.  WAD -- does the WAD identify the most likely problems customers will see and identify how the feature will enable them to diagnose and solve those problems without resorting to raising a PMR?  Have these issues been addressed in the implementation?

2.  Test and Demo -- As part of the serviceability process we're asking feature teams to test and analyze common problem paths for serviceability and demo those problem paths to someone not involved in the development of the feature (eg. L2, test team, or another development team).  
a) What problem paths were tested and demonstrated? 
b) Who did you demo to?
c) Do the people you demo'd to agree that the serviceability of the demonstrated problem scenarios is sufficient to avoid PMRs for any problems customers are likely to encounter, or that L2 should be able to quickly address those problems without need to engage L3?

3.  SVT -- SVT team is often the first team to try new features and often encounters problems setting up and using them.  Note that we're not expecting SVT to do full serviceability testing -- just to sign-off on the serviceability of the problem paths they encountered.
a) Who conducted SVT tests for this feature?
b) Do they agree that the serviceability of the problems they encountered is sufficient to avoid PMRs, or that L2 should be able to quickly address those problems without need to engage L3?  

4.  Which L2 / L3 queues will handle PMRs for this feature?  Ensure they are present in the contact reference file and in the queue contact summary, and that the respective L2/L3 teams know they are supporting it.  Ask Don Bourne if you need links or more info.

**Comment by marikaj123** [comment]
@pgunapal - Please ensure not to move the feature to GA before receiving all of the appropriate focal point approvals.

**Comment by marikaj123** [comment]
@pgunapal - Does the feature require Globalization/Translation?

**Comment by skasund** [comment]
L2 has requested STE slides for this feature.  The STE template can be found at the link below: 

https://ibm.box.com/s/1an42g7zdgmaj84w7dft0indqfgi8ffm

Please upload the completed slides to the same STE Archive BOX folder.  Thanks! 

If more time is needed to complete the slides, please provide an ETA on when they'll be available.

**Comment by dave-waddling** [comment]
**FAT Focal Approval comment:** Hi @fmhwong, I've had a look at the results from the SOE over the weekend. There are currently a pair of open defects against the TCK which will need to be resolved before I can add a FAT Focal Approval. They are: 
- [264282](https://wasrtc.hursley.ibm.com:9443/jazz/resource/itemName/com.ibm.team.workitem.WorkItem/264282)
- [264823](https://wasrtc.hursley.ibm.com:9443/jazz/resource/itemName/com.ibm.team.workitem.WorkItem/264823)

**Comment by Channyboy** [comment]
Translated files are found in master branch.
@marikaj123

**Comment by jhanders34** [comment]
The performance team identified issue #8067 where there was a significant performance difference between mp Health 1.0 and 2.0.  With the delivery of PR #8101, there is an improvement of 25% for mp Health 1.0 and 40% for mp Health 2.0.  I am proactively giving performance focal approval expecting that the PR will make it to master.

**Comment by pgunapal** [comment]
@skasund Hi, We can provide the STE for this feature, by Thursday (July 10). Can you please provide the STE Focal point approval?

**Comment by pgunapal** [comment]
@lauracowen Can I get the ID approval for this feature, the GA Blog post is in #7895

**Comment by donbourne** [comment]
Giving exception on serviceability approval.  Plan is to do the demo (2b) in the next week and address any issues found as defects.  Not expecting any issues.

**Comment by pgunapal** [comment]
**Serviceability Approval Comment** - Please answer the following questions for serviceability approval:

1.  WAD -- does the WAD identify the most likely problems customers will see and identify how the feature will enable them to diagnose and solve those problems without resorting to raising a PMR?  Have these issues been addressed in the implementation?

- Yes, the WAD includes the most likely problems the customers will see, which is usually in their own implementation of the HealthCheck interface, in their applications. Detailed messages will be logged in the messages.log.

2.  Test and Demo -- As part of the serviceability process we're asking feature teams to test and analyze common problem paths for serviceability and demo those problem paths to someone not involved in the development of the feature (eg. L2, test team, or another development team).  
a) What problem paths were tested and demonstrated?
b) Who did you demo to?
c) Do the people you demo'd to agree that the serviceability of the demonstrated problem scenarios is sufficient to avoid PMRs for any problems customers are likely to encounter, or that L2 should be able to quickly address those problems without need to engage L3?

- This will be demoed at a later date.

3.  SVT -- SVT team is often the first team to try new features and often encounters problems setting up and using them.  Note that we're not expecting SVT to do full serviceability testing -- just to sign-off on the serviceability of the problem paths they encountered.
a) Who conducted SVT tests for this feature?
- Tanya Kulik
b) Do they agree that the serviceability of the problems they encountered is sufficient to avoid PMRs, or that L2 should be able to quickly address those problems without need to engage L3?  
- Yes

4.  Which L2 / L3 queues will handle PMRs for this feature?  Ensure they are present in the contact reference file and in the queue contact summary, and that the respective L2/L3 teams know they are supporting it.  Ask Don Bourne if you need links or more info.
- WAS L3: Logging

**Comment by skasund** [comment]
@pgunapal  I'm providing the STE approval early. Continue working on the slides and provide them as soon as possible.

**Comment by chirp1** [comment]
awaiting further discussion with prashanth before approving....

**Comment by pgunapal** [comment]
@dave-waddling 
Updated the RTC defects as well (264847 and 264823). Both defects are caused by the same intermittent timing issue, during application start-up, where two of the methods do not occur in the expected order.

It only sometimes happens during application deployment (startup), where the ServletContainerInitializer.onStartup() gets called, after the HealthCheck has been completed (thinks there are no applications), resulting in a false positive health check, which fails the TCK. When the TCK is run manually, or in the continuous release builds, it passes 100% of the time. In each SOE build, a different test case fails, due to this timing issue, and only occurs intermittently, meaning in some SOE builds all the test cases pass. It occurs on z/OS builds most of the times. 

Note: This issue was always present in mpHealth-1.0, as well, however, since there were no test cases that tested this particular scenario and the TCKs were never automated for mpHealth-1.0, this issue went under the radar.

**Comment by lauracowen** [comment]
@pgunapal I can't sign off ID approval but I've checked the blog post (which is more 'dev engagement' than ID). I've asked for a little more clarification in the issue - if you can provide that, I think it's okay.

**Comment by pgunapal** [comment]
@lauracowen Provided the required clarification in Issue #7895

**Comment by donbourne** [comment]
added serviceability approval, but will continue to track this one until serviceability questions can be fully answered

**Comment by pgunapal** [comment]
@skasund Uploaded the completed STE slides in the STE Archive BOX folder.

**Comment by skasund** [comment]
@pgunapal  Thanks for the slides!

**Comment by marikaj123** [comment]
@pgunapal - Was the feature code enabled and included into the Liberty 19007 GM build?

**Comment by tevans78** [comment]
Yes, it was all delivered. Only thing missing from this epic is the FAT focal approval.

**Comment by marikaj123** [comment]
I am keeping the Target:ga to remind us that FAT needs to be complete.

**Comment by pgunapal** [comment]
Received all focal point approvals and feature was delivered in 19.0.0.7.
